_posts/2018-05-07-mmio-registers.md

```diff
@@ -0,0 +1,234 @@
+```
+
+The really cool bit about the registers system is that it prevents mistakes
+when used in a driver. For instance if you meant to write to the mode register,
+but mistakenly specified `cr` (the control register), the code would no longer
+compile.
+
+```rust
+error[E0308]: mismatched types
+   --> tock/chips/sam4l/src/usart.rs:940:13
+    |
+940 | /             Mode::MODE::NORMAL + Mode::CLK0::SET
+941 | |                 + Mode::FILTER::SET,
+    | |___________________________________^
+            expected struct `usart::Control::Register`,
+            found struct `usart::Mode::Register`
+```
+
+## SVD Files
+
+One particular pain point with the register system has been creating the
+register fields. Each peripheral has several registers, each of which can have
+up to 32 fields. For ARM microcontrollers, however, this problem has already
+been addressed with
+[SVD files](http://arm-software.github.io/CMSIS_5/SVD/html/index.html).
+
+A `System View Description` file is a formal description of the registers for
+each peripheral in a ARM microcontroller. They are created and maintained by
+the company which made the chip. A collection of SVD files for dozens of chips
+by manufacturers like STMicroelectronics, Texas Instruments, and Nordic
+Semiconductor are available in the python package
+[cmsis-svd](https://github.com/posborne/cmsis-svd).
+
+## Automatic Generation
+
+The standard format of SVD files allows them to be parsed in order to generate
+register fields. Using this, a new Tock tool, [Stefan Hölzl](https://github.com/stefanhoelzl)
+created a new tool, `svd2regs`, in (gh#877). `svd2regs` parses the SVD
+file for a microcontroller and generates the Tock register interface code for a
+specified register.
+
+We're really excited about `svd2regs` because it reduces the effort for Tock to
+support a new chip. One of the more tedious parts of adding peripheral drivers
+is the creation of its registers structures, which is now automated, reducing
+the possibility for human-error in transcription.
+
+## Comparison to `svd2rust`
+
+Tock's register interface is not the only effort to add compile-time
+checking and access control to memory-mapped registers.
+[`svd2rust`](https://github.com/japaric/svd2rust) also automatically generates
+register maps in Rust from SVD files. The
+[register system](https://docs.rs/svd2rust/0.12.1/svd2rust/) is very
+similar in capability to the Tock register interface, especially for SVDs with
+enumerated values.
+
+One notable difference between the two interfaces is that `svd2rust` relies on
+closures to guarantee compile-time checks, while Tock does not. For example,
+here is LED blink code written using both interfaces. The example is taken from
+the [STM32F042 repo](https://github.com/therealprof/stm32f042/blob/master/examples/blinky.rs).
+
+Example with the Tock register interface:
+
+```rust
+fn main() {
+    let rcc_regs: &RccRegisters = &*RCC_BASE;
+    let gpioa_regs: &GpioaRegisters = &*GPIOA_BASE;
```

Comment:
Is this cast correct? That's not typically how we use it, is it?

Reply:
It does compile. Also, that is the same way we use it in the i2c driver: https://github.com/tock/tock/blob/master/chips/sam4l/src/i2c.rs#L579

I thought we were moving to `StaticRef` where we could just do the `&*` trick without unsafe.